authenticatorbridge/src/main/java/com/bitwarden/authenticatorbridge/manager/AuthenticatorBridgeManagerImpl.kt

```diff
@@ -0,0 +1,195 @@
+                mutableSharedAccountsStateFlow.value = AccountSyncState.SyncNotEnabled
+                unBindService()
+            }
+
+            // Register callback:
+            bridgeService.safeCall { registerBridgeServiceCallback(authenticatorBridgeCallback) }
+
+            // Sync data:
+            bridgeService.safeCall { syncAccounts() }
+        }
+
+        override fun onServiceDisconnected(name: ComponentName) {
+            bridgeService = AuthenticatorBridgeServiceSafetyWrapper(service = null)
+        }
+    }
+
+    override val accountSyncStateFlow: StateFlow<AccountSyncState> =
+        mutableSharedAccountsStateFlow.asStateFlow()
+
+    init {
+        // Listen for lifecycle events
+        processLifecycleOwner.lifecycle.addObserver(
+            object : DefaultLifecycleObserver {
+                override fun onStart(owner: LifecycleOwner) {
+                    bindService()
+                }
+
+                override fun onStop(owner: LifecycleOwner) {
+                    unBindService()
+                }
+            },
+        )
+    }
+
+    private fun bindService() {
+        val intent = Intent().apply {
+            component = ComponentName(
+                connectionType.toPackageName(),
+                AUTHENTICATOR_BRIDGE_SERVICE_CLASS,
+            )
+        }
+
+        isBound = try {
+            applicationContext.bindService(
+                intent,
+                bridgeServiceConnection,
+                Context.BIND_AUTO_CREATE,
+            )
+        } catch (e: SecurityException) {
+            false
+        }
+
+        if (!isBound) {
+            mutableSharedAccountsStateFlow.value = AccountSyncState.Error
+        }
+    }
+
+    private fun onAccountsSync(data: EncryptedSharedAccountData) {
+        // Received account sync update. Decrypt with local symmetric key and update StateFlow:
+        val decryptedAccounts = symmetricKeyStorageProvider.symmetricKey?.let {
+            val decryptedAccounts = data.decrypt(it).getOrNull()?.accounts ?: return@let null
+            AccountSyncState.Success(decryptedAccounts)
+        }
+        mutableSharedAccountsStateFlow.value = decryptedAccounts
+            ?.let { AccountSyncState.Success(it.accounts) }
+            ?: AccountSyncState.Error
+    }
+
+    private fun unBindService() {
+        if (!isBound) return
+        isBound = false
+        bridgeService.safeCall { unregisterBridgeServiceCallback(authenticatorBridgeCallback) }
+        bridgeService = AuthenticatorBridgeServiceSafetyWrapper(service = null)
+        applicationContext.unbindService(bridgeServiceConnection)
+    }
+}
+
+/**
+ * Helper class for wrapping all calls to [IAuthenticatorBridgeService] around try catch.
+ *
+ * This is important because all calls to [IAuthenticatorBridgeService] can throw
+ * DeadObjectExceptions as well as RemoteExceptions.
+ */
+private class AuthenticatorBridgeServiceSafetyWrapper(
```

Comment:
Do we need a class for this? Can't we just use an extension function like:

```kotlin
    fun <T> IAuthenticatorBridgeService.safeCall(action: IAuthenticatorBridgeService.() -> T): T? =
        try {
            this?.let { action.invoke(it) }
        } catch (e: Exception) {
            null
        }
```

Reply:
mmmm, yes, I like that. My original intention with the class was to make it _impossible_ to call the service without the wrapper (by giving no access to the service directly), but I think we can trust future editors of this file to use this function and never call the service without `safeCall`.

Updated: c9fc5fb03ecc328e680d4dd05ae30e729248dc3f